### PR TITLE
poster fixes

### LINKED ISSUE #69954: Posters don't delete themselves after being used
## Reproduction:
Print a poster
Use it on wall

### PR DESCRIPTION
**Pull request by ShizCalev**
fuck poster code. whole thing needs a rewrite imo.
Fixes #69954

:cl: ShizCalev
fix: Wanted/missing posters are no longer broken.
fix: Missing posters no longer change text color after being taken down.
fix: Posters no longer fail to get put up 4 seconds after you walk away while placing them lol

/:cl:

### CODE REVIEW

File: code/game/objects/effects/contraband.dm [hunk 1]
```diff
@@ -223,23 +223,22 @@
...
-	playsound(D.loc, 'sound/items/poster_being_created.ogg', 100, TRUE)
+	flick("poster_being_set", placed_poster)
+	placed_poster.forceMove(src) //deletion of the poster is handled in poster/Exited(), so don't have to worry about P anymore.
+	playsound(src, 'sound/items/poster_being_created.ogg', 100, TRUE)
 
-	if(do_after(user, PLACE_SPEED, target=src))
-		if(!D || QDELETED(D))
-			return
+	var/atom/user_drop_location = user.drop_location() //cache this so it just falls to the ground if they move.
+	if(!do_after(user, PLACE_SPEED, placed_poster, extra_checks = CALLBACK(placed_poster, /obj/structure/sign/poster.proc/snowflake_wall_turf_check, src)))
```
Comment (timothymtorres) [inline code/game/objects/effects/contraband.dm:233]:
```suggestion
	if(!do_after(user, PLACE_SPEED, placed_poster, extra_checks = CALLBACK(placed_poster, .proc/istype, src, /turf/closed/wall)))
```

Let me know if this works.

File: code/game/objects/effects/contraband.dm [hunk 2]
```diff
@@ -223,23 +223,22 @@
...
+		return
 
-		if(iswallturf(src) && user && user.loc == temp_loc) //Let's check if everything is still there
-			D.on_placed_poster(user)
-			return
+	placed_poster.on_placed_poster(user)
 
-	to_chat(user, span_notice("The poster falls down!"))
-	D.roll_and_drop(get_turf(user))
+/obj/structure/sign/poster/proc/snowflake_wall_turf_check(atom/hopefully_still_a_wall_turf) //since turfs never get deleted but instead change type, make sure we're still being placed on a wall.
```
Comment (timothymtorres) [inline code/game/objects/effects/contraband.dm:240]: If the above callback works, this isn't needed anymore.

File: code/game/objects/effects/contraband.dm (hunk 1 above)
Comment (ShizCalev) [reply code/game/objects/effects/contraband.dm:233]: I spent almost an hour trying to figure that one out last night before opening the PR, doesn't work sadly.

File: code/game/objects/effects/contraband.dm [hunk 3]
```diff
@@ -191,10 +191,10 @@
...
 	return (!user.gloves && !HAS_TRAIT(user, TRAIT_PIERCEIMMUNE))
 
-/obj/structure/sign/poster/proc/roll_and_drop(loc)
+/obj/structure/sign/poster/proc/roll_and_drop(atom/location)
 	pixel_x = 0
 	pixel_y = 0
-	var/obj/item/poster/P = new poster_item_type(loc, src)
+	var/obj/item/poster/P = new poster_item_type(location, src) // /obj/structure/sign/poster/wanted/roll_and_drop() has some snowflake handling due to icon memes, if you make a major change to this, don't forget to update it too. <3
 	forceMove(P)
 	return P
```
[the remark this answers is not in the document]
Comment (ShizCalev) [reply code/game/objects/effects/contraband.dm:199]: Yes, the structures are stored inside the posters.

[the remark this answers is not in the document]
Comment (ShizCalev) [reply code/game/objects/effects/contraband.dm:199]: As I said, this shit needs a rewrite lol

File: code/game/objects/effects/contraband.dm (hunk 1 above)
Comment (timothymtorres) [reply code/game/objects/effects/contraband.dm:233]: Dang, kind of surprised the internal byond procs can't be used like that.  I know I asked on discord months ago about doing this for macros but Mothblocks advised me not to do it that way.

**Pull request merged**